### DISCUSSION TITLE
Steps to set up Barrier with SSL

### DISCUSSION THREAD

**Discussion by anas-abbadi09** [debauchee/barrier#2105] [Show and tell]
Here’s a detailed summary of the steps to set up Barrier with SSL on macOS (server), Windows 10 (client), and Linux machines:

---

### **Setup for macOS (Server)**
1. **Download and Install Barrier**:
   - Download the Barrier `.dmg` file from the official website.
   - Install Barrier and configure it as the **server** (under "Server Configuration").

2. **Install OpenSSL (if not already installed)**:
   - Open **Terminal** and run:
     ```bash
     brew install openssl
     ```
     (If you don’t have Homebrew, you can install it from [https://brew.sh](https://brew.sh).)

3. **Generate the SSL Certificate**:
   - Open **Terminal** and navigate to the Barrier SSL directory:
     ```bash
     cd ~/Library/Application\ Support/barrier/SSL
     ```
   - Generate the SSL certificate using OpenSSL:
     ```bash
     openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem
     ```
     This command generates the `Barrier.pem` file in the `SSL` directory with a validity of 365 days.

4. **Allow Barrier in macOS Security Settings**:
   - Go to **System Preferences > Security & Privacy > Accessibility**.
   - Add **Barrier** and grant it necessary permissions for network access and input control.

5. **Restart Barrier**:
   - Relaunch Barrier on macOS after configuring SSL.

---

### **Setup for Windows 10 (Client)**

1. **Download and Install Barrier**:
   - Download the `Barrier.exe` file and perform a fresh install.
   - Configure Barrier as the **client**.

2. **Install Git for Windows**:
   - Download and install Git for Windows, using standard configurations.

3. **Generate the SSL Certificate**:
   - Open **Git Bash** and navigate to the Barrier SSL directory:
     ```bash
     cd "/c/Users/YourUsername/AppData/Local/Barrier/SSL"
     ```
   - Run the following OpenSSL command:
     ```bash
     openssl req -x509 -nodes -days 365 -subj //CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem
     ```
     Ensure you use **double forward slashes** in the `-subj` argument to avoid POSIX path issues.

4. **Restart Barrier Client**:
   - After generating the SSL certificate, click the **Reload** button on the Barrier client interface.

5. **Confirm SSL Fingerprint**:
   - When connecting to the macOS server, Barrier on Windows will prompt you to confirm the fingerprint of the certificate. Accept the fingerprint to establish the connection.

---

### **Setup for Linux (Client)**

1. **Download and Install Barrier**:
   - Install Barrier using your package manager (for example, on Ubuntu):
     ```bash
     sudo apt install barrier
     ```
   - Configure Barrier as the **client**.

2. **Generate the SSL Certificate**:
   - Open a terminal and navigate to the SSL directory:
     ```bash
     cd ~/.config/barrier/SSL
     ```
   - Generate the SSL certificate using OpenSSL:
     ```bash
     openssl req -x509 -nodes -days 365 -subj /CN=Barrier -newkey rsa:4096 -keyout Barrier.pem -out Barrier.pem
     ```

3. **Disable Wayland (if needed)**:
   - On some Linux distributions, Barrier may work better with Xorg. If you're using **Wayland**, try switching to **Xorg**:
     - Open and edit the custom configuration file:
       ```bash
       sudo nano /etc/gdm3/custom.conf
       ```
     - Uncomment `#WaylandEnable=false` to disable Wayland:
       ```bash
       WaylandEnable=false
       ```
     - Save and restart GDM:
       ```bash
       sudo systemctl restart gdm3
       ```

4. **Restart Barrier Client**:
   - Relaunch Barrier on your Linux client.

5. **Disable Wi-Fi Power Saving (Optional)**:
   - If you experience connection lags on Linux due to power saving:
     - Open the configuration file:
       ```bash
       sudo nano /etc/NetworkManager/conf.d/default-wifi-powersave-on.conf
       ```
     - Change `wifi.powersave = 3` to `wifi.powersave = 2`.
     - Save and restart NetworkManager:
       ```bash
       sudo systemctl restart NetworkManager
       ```

6. **Confirm SSL Fingerprint**:
   - Barrier on Linux will prompt you to confirm the SSL certificate fingerprint from the macOS server. Accept it to establish the secure connection.

**Comment by blogdron** [comment]
In Mint Zena (and maybe others) need store `Barrier.pem` file in  `~/.local/share/barrier/SSL/`

**Reply by nbolton** [reply]
Barrier is no longer in development. Check out Deskflow (upstream) or Input Leap (fork).

https://github.com/deskflow/deskflow
https://github.com/input-leap/input-leap

If this is still an issue in those projects, we would appreciate a cross-post of this issue.

**Reply by blogdron** [reply]
No no, its just notice (about specific file location), for someone maybe usefull, Mint have Barrier in package repos and Barrier working very fine now, in another PC I use Deskflow. Yes Im connect (client Mint) Barrier to (server Debian) Deskflow :D  Just becouse Deskflow in Mint only in flatpak.  Its not issue, just notice.  For those who stumble here by accident, a barrier makes sense, for example, for systems that can no longer be updated and for systems that prohibit installation outside the repository. At the protocol level, both the parent project and its forks are compatible. So, if a project is no longer being developed, that doesn't mean it doesn't work or has become worse. It's still the same code. And thanks to those who wrote it. When everything is broken, then I will think about switching to a fork. That's all for me :)

**Reply by nbolton** [reply]
Indeed, if you're looking for improvements, you should look upstream to Deskflow which is where all changes are happening now (Barrier is an unmaintained downstream fork). Input Leap (another downstream fork) hasn't had a release in almost a year, but some people still use it as it's got some stuff that hasn't yet been ported upstream to Deskflow.
